README.md
# dhooks-lite

Another simple class wrapper for interacting with Discord webhooks.

[![release](https://img.shields.io/pypi/v/dhooks-lite?label=release)](https://pypi.org/project/dhooks-lite/)
[![python](https://img.shields.io/pypi/pyversions/dhooks-lite)](https://pypi.org/project/dhooks-lite/)
[![pipeline status](https://gitlab.com/ErikKalkoken/dhooks-lite/badges/master/pipeline.svg)](https://gitlab.com/ErikKalkoken/dhooks-lite/-/commits/master)
[![codecov](https://codecov.io/gl/ErikKalkoken/dhooks-lite/branch/master/graph/badge.svg?token=9YNL3HEJ5D)](https://codecov.io/gl/ErikKalkoken/dhooks-lite)
[![Documentation Status](https://readthedocs.org/projects/dhooks-lite/badge/?version=latest)](https://dhooks-lite.readthedocs.io/en/latest/?badge=latest)
[![license](https://img.shields.io/badge/license-MIT-green)](https://gitlab.com/ErikKalkoken/dhooks-lite/-/blob/master/LICENSE)
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![chat](https://img.shields.io/discord/790364535294132234)](https://discord.gg/zmh52wnfvM)

## Contents

- [Overview](#overview)
- [Functionality](#functionality)
- [Examples](#examples)
- [Installation](#installation)
- [Documentation](#documentation)
- [Contribution](#contribution)

## Overview

**dhooks-lite** is a library with a set of classes for interacting with Discord webhooks written in Python 3.

This library aims to differentiate itself from similar libraries with the following properties:

- is fully tested
- simple to use (only one way of doing things, same name of attributes and objects as in the [official Discord documentation](https://discordapp.com/developers/docs/resources/webhook#execute-webhook))
- has logging
- requests are automatically retried and have sensible timeouts
- works with older Python versions

## Functionality

This library provides following functionality:

- Posting messages in Discord channels via webhooks (synchronous calls only)
- Attaching Embeds to messages
- Retrieve send reports and from Discord
- Retrieve HTTP status and headers from Discord, e.g. for implementing rate limit handling

## Examples

Here are some examples on how to use dhooks-lite in your Python scripts.

Note that you also find the source code of all examples in the `/examples` folder of this repo.

### Hello World

Minimal example for posting a message.

```python
from dhooks_lite import Webhook

hook = Webhook(DISCORD_WEBHOOK_URL)
hook.execute('Hello, World!')
```

![example1](https://i.imgur.com/t3mxMAJ.png)

### Posting with custom avatar

In this example we are setting username and avatar.

```python
from dhooks_lite import Webhook

hook = Webhook(
    DISCORD_WEBHOOK_URL,
    username='Bruce Wayne',
    avatar_url='https://i.imgur.com/thK8erv.png'
)
hook.execute('I am Batman!')
```

![example2](https://i.imgur.com/mseg2Yx.png)

### Complete example with embeds

Finally, here is an example for posting a message with two embeds and using all available features (shortened):

```python
import datetime
from dhooks_lite import Webhook, Embed, Footer, Image, Thumbnail, Author, Field

hook = Webhook(DISCORD_WEBHOOK_URL)
e1 = Embed(
    description='Only a few years ago, scientists stumbled upon an electrical current of cosmic proportions.(...)',
    title='Universe\'s highest electric current found',
    url='https://www.newscientist.com/article/mg21028174-900-universes-highest-electric-current-found/',
    timestamp=datetime.datetime.now(),
    color=0x5CDBF0,
    footer=Footer(
        'Science Department',
        'https://i.imgur.com/Bgsv04h.png'
    ),
    image=Image('https://i.imgur.com/eis1Y0P.jpg'),
    thumbnail=Thumbnail('https://i.imgur.com/2A4k28x.jpg'),
    author=Author(
        'John Scientist',
        url='https://en.wikipedia.org/wiki/Albert_Einstein',
        icon_url='https://i.imgur.com/1JoHDw1.png'
    ),
    fields=[
        Field('1st Measurement', 'Failed'),
        Field('2nd Measurement', 'Succeeded')
    ]
)
e2 = Embed(description="TOP SECRET - Do not distribute!")

hook.execute(
    'Checkout this new report from the science department:',
    username='Bruce Wayne',
    avatar_url='https://i.imgur.com/thK8erv.png',
    embeds=[e1, e2],
    wait_for_response=True
)
```

![example2](https://i.imgur.com/RoWBh2n.png)

## Installation

You can install this library directly from PyPI:

```bash
pip install dhooks-lite
```

## Documentation

For a full documentation of all classes please see the official docs [here](https://dhooks-lite.readthedocs.io/en/latest/).

## Contribution

We welcome any contribution!

If you found a bug or have a feature request please raise an issue.

If you want to help further improve this library please feel free to issue a merge request. Please adhere to the following requirements in your change:

- Code should be compliant with [PEP8](https://www.python.org/dev/peps/pep-0008/)
- Full overage by unit tests
- All classes should be immutable
- All classes and their public methods must have docstring documentation
- All changes must be documented in the Change Log


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - yyyy-mm-dd

## [1.1.0] - 2023-07-16

### Added

- Add support for Python 3.11
- isort linting
- Added pylint linter to CI pipeline

### Changed

- Migrated build process to PEP 621
- Migrated docs to alabama theme

## [1.0.0] - 2023-04-18

### Added

- Support for Python 3.9 & 3.10
- Improved type hints

## [0.6.1] - 2021-01-03

### Fixed

- Install aborts without requests: `ModuleNotFoundError: No module named 'requests'`

## [0.6.0] - 2021-01-02

### Added

- Ability to set an user agent for all requests
- Ability to set max retries to any value and disable it

### Changed

- Remove support for Python 3.5

## [0.5.1] - 2020-12-30

### Fixed

- Did not return a WebhookResponse object for http errors

## [0.5.0] - 2020-07-03

### Added

- Added new ability to create an Embed from a dict
- Added missing type hints
- Added API documentation for `from_dict()` and `asdict()`

### Changes

- Removed support for Python 3.4
- Renamed `to_dict()` to `asdict()`

## [0.4.1] - 2020-06-15

### Fixed

- Wrong PyPI upload format

## [0.4.0] - 2020-06-14

### Added

- Automatic retries on retry-able HTTP errors with exponential backoff
- Timeout for requests
- Included response headers in debug log

## [0.3.2] - 2020-03-31

### Fixed

- Removed deprecated attributes, added missing test cases

## [0.3.1] - 2019-12-07

### Fixed

- pip install error when "requests" module was not already present

## [0.3.0] - 2019-12-04

### Added

- Webhooks returns HTTP status code and headers. This allows the implementation of rate limit handling.

### Changed

- Webhook.execute() now returns a response object instead of the send report. The response objects contains the HTTP status code, headers and content of the response from Discord (e.g. send report)

- Webhook.execute() wil no longer throw an exception on non 2xx HTTP codes. Instead the application code needs to investigate the status code that is returned. It will still through exceptions on all network related issues.

## [0.2.0] - 2019-11-12

### Added

- Initial


tests/test_embed.py
from datetime import datetime
from unittest import TestCase

from dhooks_lite.embed import Author, Embed, Field, Footer, Image, Thumbnail
from tests.utils import set_test_logger

MODULE_PATH = "dhooks_lite.embed"
logger = set_test_logger(MODULE_PATH, __file__)


class TestEmbedObjectComparing(TestCase):
    def setUp(self):
        self.x1 = Author("Bruce", "url-1")
        self.x2 = Author("Bruce", "url-1")
        self.y1 = Author("Bruce", "url-2")
        self.y2 = Author("Clark", "url-1")
        self.z = Author("Clark", "url-2")

    def test_objects_are_equal(self):
        self.assertEqual(self.x1, self.x1)
        self.assertEqual(self.x1, self.x2)

    def test_objects_are_not_equal(self):
        self.assertNotEqual(self.x1, self.y1)
        self.assertNotEqual(self.x1, self.y2)
        self.assertNotEqual(self.x1, self.z)
        self.assertNotEqual(self.x1, Footer("Bruce", "url-1"))


class TestEmbedObject(TestCase):
    def test_fromdict(self):
        field = Field(name="fruit", value="orange", inline=False)
        field_dict = field.asdict()
        field2 = Field.from_dict(field_dict)
        self.assertEqual(field, field2)


class TestAuthor(TestCase):
    def test_detect_missing_params_on_create(self):
        with self.assertRaises(ValueError):
            Author(None)

    def test_create_with_name_only(self):
        x = Author("Bruce Wayne")
        self.assertEqual(x.name, "Bruce Wayne")
        self.assertDictEqual(x.asdict(), {"name": "Bruce Wayne"})

    def test_create_with_all_params(self):
        x = Author("Bruce Wayne", url="url-1", icon_url="url-2", proxy_icon_url="url-3")
        self.assertEqual(x.name, "Bruce Wayne")
        self.assertEqual(x.url, "url-1")
        self.assertEqual(x.icon_url, "url-2")
        self.assertEqual(x.proxy_icon_url, "url-3")
        self.assertDictEqual(
            x.asdict(),
            {
                "name": "Bruce Wayne",
                "url": "url-1",
                "icon_url": "url-2",
                "proxy_icon_url": "url-3",
            },
        )


class TestField(TestCase):
    def test_detect_missing_params_on_create(self):
        with self.assertRaises(ValueError):
            Field(name=None, value=None)

    def test_detects_name_limit(self):
        large_string = "x" * 257
        with self.assertRaises(ValueError):
            Field(large_string, value="Batman")

    def test_detects_value_limit(self):
        large_string = "x" * 1025
        with self.assertRaises(ValueError):
            Field(name="Bruce Wayne", value=large_string)

    def test_detect_missing_value(self):
        with self.assertRaises(ValueError):
            Field(name="Bruce Wayne", value=None)

    def test_detect_missing_name(self):
        with self.assertRaises(ValueError):
            Field(name=None, value="Batman")

    def test_create_with_name_and_value_only(self):
        x = Field("fruit", "orange")
        self.assertEqual(x.name, "fruit")
        self.assertEqual(x.value, "orange")
        self.assertEqual(x.inline, True)
        self.assertDictEqual(
            x.asdict(), {"name": "fruit", "value": "orange", "inline": True}
        )

    def test_create_with_all_params(self):
        x = Field(name="fruit", value="orange", inline=False)
        self.assertEqual(x.name, "fruit")
        self.assertEqual(x.value, "orange")
        self.assertEqual(x.inline, False)
        self.assertDictEqual(
            x.asdict(), {"name": "fruit", "value": "orange", "inline": False}
        )

    def test_detect_invalid_inline_type(self):
        with self.assertRaises(TypeError):
            Field(name="fruit", value="orange", inline=int(5))


class TestFooter(TestCase):
    def test_detect_missing_params_on_create(self):
        with self.assertRaises(ValueError):
            Footer(None)

    def test_detects_wrong_type_inline(self):
        with self.assertRaises(TypeError):
            Footer("Justice League", inline=int(1))

    def test_create_with_name_only(self):
        x = Footer("Justice League")
        self.assertEqual(x.text, "Justice League")
        self.assertDictEqual(x.asdict(), {"text": "Justice League"})

    def test_create_with_all_params(self):
        x = Footer("Justice League", icon_url="url-1", proxy_icon_url="url-2")
        self.assertEqual(x.text, "Justice League")
        self.assertEqual(x.icon_url, "url-1")
        self.assertEqual(x.proxy_icon_url, "url-2")
        self.assertDictEqual(
            x.asdict(),
            {"text": "Justice League", "icon_url": "url-1", "proxy_icon_url": "url-2"},
        )


class TestImage(TestCase):
    def test_detect_missing_params_on_create(self):
        with self.assertRaises(ValueError):
            Image(None)

    def test_create_with_url_only(self):
        x = Image("my-url")
        self.assertEqual(x.url, "my-url")
        self.assertDictEqual(x.asdict(), {"url": "my-url"})

    def test_create_with_all_params(self):
        x = Image(url="url-1", proxy_url="url-2", width=500, height=400)
        self.assertEqual(x.url, "url-1")
        self.assertEqual(x.proxy_url, "url-2")
        self.assertEqual(x.width, 500)
        self.assertEqual(x.height, 400)
        self.assertDictEqual(
            x.asdict(),
            {"url": "url-1", "proxy_url": "url-2", "width": 500, "height": 400},
        )

    def test_detect_invalid_width(self):
        with self.assertRaises(ValueError):
            Image("my-url", width=-5)

    def test_detect_invalid_height(self):
        with self.assertRaises(ValueError):
            Image("my-url", height=-5)


class TestEmbed(TestCase):
    def setUp(self) -> None:
        self.now = datetime.utcnow()

    def test_create_with_description_only(self):
        x = Embed(description="They said the age of heroes would never come again.")
        self.assertEqual(
            x.description, "They said the age of heroes would never come again."
        )
        self.assertEqual(x.type, "rich")
        self.assertDictEqual(
            x.asdict(),
            {
                "type": "rich",
                "description": "They said the age of heroes would never come again.",
            },
        )

    def test_create_with_full_params(self):
        obj = Embed(
            title="Justice League",
            description="They said the age of heroes would never come again.",
            url="url-1",
            timestamp=self.now,
            color=0x5CDBF0,
            footer=Footer("TOP SECRET", "url-2", "url-11"),
            image=Image("url-3", "url-4", height=200, width=150),
            thumbnail=Thumbnail("url-5", "url-6", height=100, width=80),
            author=Author("Bruce Wayne", "url-8", "url-9"),
            fields=[Field("fruit", "orange", False), Field("vegetable", "onion", True)],
        )
        self.assertEqual(obj.title, "Justice League")
        self.assertEqual(
            obj.description, "They said the age of heroes would never come again."
        )
        self.assertEqual(obj.type, "rich")
        self.assertEqual(obj.url, "url-1")
        self.assertEqual(obj.timestamp, self.now)
        self.assertEqual(obj.color, 0x5CDBF0)
        self.assertEqual(obj.footer, Footer("TOP SECRET", "url-2", "url-11"))
        self.assertEqual(obj.image, Image("url-3", "url-4", height=200, width=150))
        self.assertEqual(
            obj.thumbnail, Thumbnail("url-5", "url-6", height=100, width=80)
        )
        self.assertEqual(obj.author, Author("Bruce Wayne", "url-8", "url-9"))
        self.assertEqual(
            obj.fields,
            [Field("fruit", "orange", False), Field("vegetable", "onion", True)],
        )

        self.maxDiff = None
        self.assertDictEqual(
            obj.asdict(),
            {
                "title": "Justice League",
                "type": "rich",
                "description": "They said the age of heroes would never come again.",
                "url": "url-1",
                "timestamp": self.now,
                "color": 0x5CDBF0,
                "image": {
                    "url": "url-3",
                    "proxy_url": "url-4",
                    "height": 200,
                    "width": 150,
                },
                "thumbnail": {
                    "url": "url-5",
                    "proxy_url": "url-6",
                    "height": 100,
                    "width": 80,
                },
                "footer": {
                    "text": "TOP SECRET",
                    "icon_url": "url-2",
                    "proxy_icon_url": "url-11",
                },
                "author": {"name": "Bruce Wayne", "url": "url-8", "icon_url": "url-9"},
                "fields": [
                    {"name": "fruit", "value": "orange", "inline": False},
                    {"name": "vegetable", "value": "onion", "inline": True},
                ],
            },
        )

    def test_from_dict_minimal(self):
        embed = Embed(description="Dummy")
        embed_dict = embed.asdict()
        embed_2 = Embed.from_dict(embed_dict)
        self.assertEqual(embed, embed_2)

    def test_from_dict_full(self):
        embed = Embed(
            title="Justice League",
            description="They said the age of heroes would never come again.",
            url="url-1",
            timestamp=self.now,
            color=0x5CDBF0,
            footer=Footer("TOP SECRET", "url-2", "url-11"),
            image=Image("url-3", "url-4", height=200, width=150),
            thumbnail=Thumbnail("url-5", "url-6", height=100, width=80),
            author=Author("Bruce Wayne", "url-8", "url-9"),
            fields=[Field("fruit", "orange", False), Field("vegetable", "onion", True)],
        )
        embed_dict = embed.asdict()
        embed_2 = Embed.from_dict(embed_dict)
        self.assertEqual(embed, embed_2)

    def test_detects_wrong_type_timestamp(self):
        with self.assertRaises(TypeError):
            Embed(timestamp=int(1))

    def test_detects_wrong_type_footer(self):
        with self.assertRaises(TypeError):
            Embed(footer=int(1))

    def test_detects_wrong_type_image(self):
        with self.assertRaises(TypeError):
            Embed(image=int(1))

    def test_detects_wrong_type_thumbnail(self):
        with self.assertRaises(TypeError):
            Embed(thumbnail=int(1))

    def test_detects_wrong_type_author(self):
        with self.assertRaises(TypeError):
            Embed(author=int(1))

    def test_detects_wrong_type_fields_list(self):
        with self.assertRaises(TypeError):
            Embed(fields=int(1))

    def test_detects_wrong_type_fields_content(self):
        with self.assertRaises(TypeError):
            Embed(fields=[int(1), Field("x", 1)])

    def test_detects_max_embed_limit(self):
        description = "x" * 2000
        fields = list()
        for x in range(5):
            fields.append(Field(name="name" + str(x), value="value" + "x" * 1000))
        with self.assertRaises(ValueError):
            x = Embed(description=description, fields=fields)

    def test_detects_max_description_limit(self):
        large_string = "x" * 2049
        with self.assertRaises(ValueError):
            Embed(description=large_string)

    def test_detects_max_title_limit(self):
        large_string = "x" * 257
        with self.assertRaises(ValueError):
            Embed(title=large_string)

    def test_detects_max_fields_limit(self):
        fields = list()
        for x in range(26):
            fields.append(Field(name="name {}".format(x), value="value {}".format(x)))
        with self.assertRaises(ValueError):
            x = Embed(fields=fields)


tests/test_serializers.py
import json
from datetime import datetime
from unittest import TestCase

from dhooks_lite.embed import Author
from dhooks_lite.serializers import JsonDateTimeEncoder
from tests.utils import set_test_logger

MODULE_PATH = "dhooks_lite.serializers"
logger = set_test_logger(MODULE_PATH, __file__)


class TestJsonDateTimeEncoder(TestCase):
    def test_should_encode_datetime_without_error(self):
        # given
        my_date = datetime(2020, 8, 3, 16, 5, 0)
        my_dict = {"alpha": "dummy", "bravo": my_date}
        # when
        result = json.loads(json.dumps(my_dict, cls=JsonDateTimeEncoder))
        # then
        self.assertEqual(result, {"alpha": "dummy", "bravo": "2020-08-03T16:05:00"})

    def test_should_raise_error_for_unknown_type(self):
        # given
        my_dict = {"unknown": Author("dummy")}
        # when/then
        with self.assertRaises(TypeError):
            json.loads(json.dumps(my_dict, cls=JsonDateTimeEncoder))


dhooks_lite/embed.py
"""Discord Embed."""

import json
import logging
from datetime import datetime
from typing import Any, List, Optional, Union, get_type_hints

from .serializers import JsonDateTimeEncoder

logger = logging.getLogger(__name__)


class _EmbedObject:
    """base class for all Embed objects"""

    def __eq__(self, other: Any) -> bool:
        """enables comparing all objects by value, including nested objects"""
        if not isinstance(other, type(self)):
            return False
        return all(
            self.__dict__[key1] == other.__dict__[key2]
            for key1, key2 in zip(self.__dict__.keys(), other.__dict__.keys())
        )

    def __ne__(self, other: Any) -> bool:
        """enables comparing all objects by value, including nested objects"""
        return not self.__eq__(other)

    def asdict(self) -> dict:
        """returns a dict representation of this object

        will not include properties that are None
        """
        arr = {}
        for key, value in self.__dict__.items():
            if value is not None:
                if isinstance(value, list):
                    v_list = []
                    for elem in value:
                        if isinstance(elem, _EmbedObject):
                            v_list.append(elem.asdict())
                        else:
                            raise NotImplementedError()
                    arr[key[1:]] = v_list
                else:
                    if isinstance(value, _EmbedObject):
                        arr[key[1:]] = value.asdict()
                    else:
                        arr[key[1:]] = value
        return arr

    @classmethod
    def from_dict(cls, obj_dict: dict):
        """creates a new object from the given dict"""
        args = {}
        for param_name, param_type in get_type_hints(cls.__init__).items():
            if param_name in obj_dict and param_name != "return":
                if hasattr(param_type, "__origin__") and param_type.__origin__ == Union:
                    param_type = param_type.__args__[0]
                try:
                    origin_type = param_type.__origin__
                except AttributeError:
                    origin_type = param_type

                if issubclass(origin_type, list):
                    my_type = list(param_type.__args__).pop()
                    value = [my_type(**obj) for obj in obj_dict[param_name]]

                elif issubclass(origin_type, _EmbedObject):
                    value = param_type.from_dict(obj_dict[param_name])

                else:
                    value = obj_dict[param_name]

                args[param_name] = value

        return cls(**args)


class Author(_EmbedObject):
    """Author in an Embed"""

    def __init__(
        self,
        name: str,
        url: Optional[str] = None,
        icon_url: Optional[str] = None,
        proxy_icon_url: Optional[str] = None,
    ):
        if not name:
            raise ValueError("name can not be None")

        self._name = str(name)
        self._url = str(url) if url else None
        self._icon_url = str(icon_url) if icon_url else None
        self._proxy_icon_url = str(proxy_icon_url) if proxy_icon_url else None

    @property
    def name(self) -> str:
        """Return author name."""
        return self._name

    @property
    def url(self) -> Optional[str]:
        """Return author URL."""
        return self._url

    @property
    def icon_url(self) -> Optional[str]:
        """Return author's icon URL."""
        return self._icon_url

    @property
    def proxy_icon_url(self) -> Optional[str]:
        """Return author's proxy icon URL."""
        return self._proxy_icon_url


class Field(_EmbedObject):
    """Field in an Embed"""

    MAX_CHARACTERS_NAME = 256
    MAX_CHARACTERS_VALUE = 1024

    def __init__(self, name: str, value: str, inline: bool = True) -> None:
        if not name:
            raise ValueError("name can not be None")
        if not value:
            raise ValueError("value can not be None")
        if not isinstance(inline, bool):
            raise TypeError("inline must be of type bool")

        name = str(name)
        if len(name) > self.MAX_CHARACTERS_NAME:
            raise ValueError(
                f"name can not exceed {self.MAX_CHARACTERS_NAME} characters"
            )
        value = str(value)
        if len(value) > self.MAX_CHARACTERS_VALUE:
            raise ValueError(
                f"value can not exceed {self.MAX_CHARACTERS_VALUE} characters"
            )

        self._name = name
        self._value = value
        self._inline = inline

    @property
    def name(self) -> str:
        """Return field name."""
        return self._name

    @property
    def value(self) -> str:
        """Return field value."""
        return self._value

    @property
    def inline(self) -> Optional[bool]:
        """Return field inline."""
        return self._inline


class Footer(_EmbedObject):
    """Footer in an Embed"""

    def __init__(
        self,
        text: str,
        icon_url: Optional[str] = None,
        proxy_icon_url: Optional[str] = None,
    ) -> None:
        if not text:
            raise ValueError("text can not be None")

        self._text = str(text)
        self._icon_url = str(icon_url) if icon_url else None
        self._proxy_icon_url = str(proxy_icon_url) if proxy_icon_url else None

    @property
    def text(self) -> str:
        """Return Footer text."""
        return self._text

    @property
    def icon_url(self) -> Optional[str]:
        """Return footer's icon URL."""
        return self._icon_url

    @property
    def proxy_icon_url(self) -> Optional[str]:
        """Return footer's proxy icon URL."""
        return self._proxy_icon_url


class Image(_EmbedObject):
    """Image in an Embed"""

    def __init__(
        self,
        url: str,
        proxy_url: Optional[str] = None,
        height: Optional[int] = None,
        width: Optional[int] = None,
    ) -> None:
        if not url:
            raise ValueError("url can not be None")
        if width and width <= 0:
            raise ValueError("width must be > 0")
        if height and height <= 0:
            raise ValueError("height must be > 0")

        self._url = str(url)
        self._proxy_url = str(proxy_url) if proxy_url else None
        self._height = int(height) if height else None
        self._width = int(width) if width else None

    @property
    def url(self) -> str:
        """Return image URL."""
        return self._url

    @property
    def proxy_url(self) -> Optional[str]:
        """Return image's proxy URL."""
        return self._proxy_url

    @property
    def height(self) -> Optional[int]:
        """Return image height."""
        return self._height

    @property
    def width(self) -> Optional[int]:
        """Return image width."""
        return self._width


class Thumbnail(Image):
    """Thumbnail in an Embed"""


class Embed(_EmbedObject):
    """Embedded content for a message"""

    # pylint: disable=too-many-instance-attributes

    MAX_CHARACTERS = 6000
    MAX_TITLE = 256
    MAX_DESCRIPTION = 2048
    MAX_FIELDS = 25

    def __init__(
        self,
        description: Optional[str] = None,
        title: Optional[str] = None,
        url: Optional[str] = None,
        timestamp: Optional[datetime] = None,
        color: Optional[int] = None,
        footer: Optional[Footer] = None,
        image: Optional[Image] = None,
        thumbnail: Optional[Thumbnail] = None,
        author: Optional[Author] = None,
        fields: Optional[List[Field]] = None,
    ) -> None:
        """Initialize an Embed object

        Args:
            description: message text for this embed
            title: title of embed
            url: url of embed
            timestamp: timestamp of embed content
            color: color code of the embed
            footer: footer information
            image: image within embed
            thumbnail: thumbnail for this embed
            author: author information
            fields: fields information

        Exceptions:
            TypeException: when passing variables of wrong type
            ValueException: when embed size exceeds hard limit

        """
        if timestamp and not isinstance(timestamp, datetime):
            raise TypeError("timestamp must be a datetime object")

        if footer and not isinstance(footer, Footer):
            raise TypeError("footer must be a Footer object")

        if image and not isinstance(image, Image):
            raise TypeError("image must be an Image object")

        if thumbnail and not isinstance(thumbnail, Thumbnail):
            raise TypeError("thumbnail must be a Thumbnail object")

        if author and not isinstance(author, Author):
            raise TypeError("author must be a Author object")

        if fields and not isinstance(fields, list):
            raise TypeError("fields must be a list")

        if fields:
            if len(fields) > self.MAX_FIELDS:
                raise ValueError(f"Fields can not exceed {self.MAX_FIELDS} objects")
            for field in fields:
                if not isinstance(field, Field):
                    raise TypeError("all elements in fields must be a Field")

        if description and len(description) > self.MAX_DESCRIPTION:
            raise ValueError(
                f"description exceeds max length of {self.MAX_DESCRIPTION} characters"
            )

        if title and len(title) > self.MAX_TITLE:
            raise ValueError(f"title exceeds max length of {self.MAX_TITLE} characters")

        self._title = str(title) if title else None
        self._type = "rich"
        self._description = str(description) if description else None
        self._url = str(url) if url else None
        self._timestamp = timestamp
        self._color = int(color) if color else None
        self._footer = footer
        self._image = image
        self._thumbnail = thumbnail
        self._author = author
        self._fields = fields

        d_json = json.dumps(self.asdict(), cls=JsonDateTimeEncoder)
        if len(d_json) > self.MAX_CHARACTERS:
            limit = len(d_json) - self.MAX_CHARACTERS
            raise ValueError(
                f"Embed exceeds maximum allowed char size of {self.MAX_CHARACTERS} "
                f"by {limit}"
            )

    @property
    def description(self) -> Optional[str]:
        """Return embed's description."""
        return self._description

    @property
    def title(self) -> Optional[str]:
        """Return embed's title or None."""
        return self._title

    @property
    def type(self) -> Optional[str]:
        """Return embed's type or None."""
        return self._type

    @property
    def url(self) -> Optional[str]:
        """Return embed's URL or None."""
        return self._url

    @property
    def timestamp(self) -> Optional[datetime]:
        """Return embed's timestamp or None."""
        return self._timestamp

    @property
    def color(self) -> Optional[int]:
        """Return embed's color or None."""
        return self._color

    @property
    def footer(self) -> Optional[Footer]:
        """Return embed's footer or None."""
        return self._footer

    @property
    def image(self) -> Optional[Image]:
        """Return embed's image or None."""
        return self._image

    @property
    def thumbnail(self) -> Optional[Thumbnail]:
        """Return embed's thumbnail or None."""
        return self._thumbnail

    @property
    def author(self) -> Optional[Author]:
        """Return embed's author or None."""
        return self._author

    @property
    def fields(self) -> Optional[List[Field]]:
        """Return embed's fields or None."""
        return self._fields


examples/examples.py
import datetime
import logging
import os

from dhooks_lite import Author, Embed, Field, Footer, Image, Thumbnail, Webhook

logging.basicConfig(format="%(levelname)s: %(message)s", level=logging.INFO)

if "DISCORD_WEBHOOK_URL" not in os.environ:
    raise ValueError('mandatory environment variable "DISCORD_WEBHOOK_URL" not set')
else:
    DISCORD_WEBHOOK_URL = os.environ["DISCORD_WEBHOOK_URL"]


# Minimal example: Hello World
hook = Webhook(DISCORD_WEBHOOK_URL)
hook.execute("Hello, World!")

# Set username and avatar URL
hook = Webhook(
    DISCORD_WEBHOOK_URL,
    username="Bruce Wayne",
    avatar_url="https://i.imgur.com/thK8erv.png",
)
hook.execute("I am Batman!")

# Minmal embed example
hook = Webhook(DISCORD_WEBHOOK_URL)
e = Embed(description="Simple Embed example")
hook.execute(embeds=[e])


# Example with two embeds and all parameters
hook = Webhook(DISCORD_WEBHOOK_URL)
e1 = Embed(
    description="Only a few years ago, scientists stumbled upon an electrical current of cosmic proportions: 10^18 amps, or roughly one trillion lightning bolts. The lightning is thought to originate from an enormous black hole in the center of the galaxy, which has a core that is supposedly a “huge cosmic jet.” Apparently, the black hole’s huge magnetic field allows it to fire up this lightning bolt through gas and dust to a distance of over one hundred and fifty thousand light years away. And we thought that our galaxy was big, this single lightning bolt is one and a half times the size of it.",  # noqa
    title="Universe's highest electric current found",
    url="https://www.newscientist.com/article/mg21028174-900-universes-highest-electric-current-found/",  # noqa
    timestamp=datetime.datetime.utcnow(),
    color=0x5CDBF0,
    footer=Footer("Science Department", "https://i.imgur.com/Bgsv04h.png"),
    image=Image("https://i.imgur.com/eis1Y0P.jpg"),
    thumbnail=Thumbnail("https://i.imgur.com/2A4k28x.jpg"),
    author=Author(
        "John Scientist",
        url="https://en.wikipedia.org/wiki/Albert_Einstein",
        icon_url="https://i.imgur.com/1JoHDw1.png",
    ),
    fields=[Field("1st Measurement", "Failed"), Field("2nd Measurement", "Succeeded")],
)
e2 = Embed(description="TOP SECRET - Do not distribute!")

response = hook.execute(
    "Checkout this new report from the science department:",
    username="Bruce Wayne",
    avatar_url="https://i.imgur.com/thK8erv.png",
    embeds=[e1, e2],
    wait_for_response=True,
)
print(response.headers)
print()
print(response.content)
